File: projects/06-object-store/src/erasure/gf256.rs
//! GF(2⁸) arithmetic — the number system every shard byte lives in.
//!
//! Ordinary integer math overflows and rounds. Bytes need a field that is
//! **closed over `0..255`**, has inverses for every nonzero element, and where
//! add = subtract = XOR. That field is GF(2⁸) with reduction polynomial
//! **`0x11D`** (AES / RAID-6 convention).
//!
//! ## Layout
//!
//! 1. Precompute `log[256]` / `antilog[256]` once via repeated [`Gf256::xtime`] from 1.
//! 2. [`Gf256::mul`] is `antilog[(log[a] + log[b]) mod 255]` (with `0` short-circuit).
//! 3. [`Gf256::inv`] / [`Gf256::div`] from the tables (`a⁻¹ = antilog[255 − log[a]]`).
//!
//! Bit-exact against the hand-worked products in
//! [`docs/12-how-erasure-coding-works.md`](../../../docs/12-how-erasure-coding-works.md)
//! §3–§4 — a wrong table silently corrupts every shard.
//!
//! `add` is free (XOR).

use super::ErasureError;

/// GF(2⁸) with AES/RAID-6 reduction polynomial `0x11D`.
///
/// Hold one of these (or a `&'static`) and pass it into the RS / LRC encoders.
#[derive(Debug, Clone)]
pub struct Gf256 {
    /// `log[x] = i` such that `2ⁱ ≡ x` (undefined / unused at `log[0]`).
    pub log: [u8; 256],
    /// `antilog[i] = 2ⁱ` for `i` in `0..255` (index 255 mirrors 0: `2²⁵⁵ = 2⁰ = 1`).
    pub antilog: [u8; 256],
}

impl Gf256 {
    /// Irreducible polynomial used by AES and Linux RAID-6 (`x⁸ + x⁴ + x³ + x² + 1`).
    pub const REDUCTION_POLY: u16 = 0x11D;

    /// Build log/antilog tables. Call once; reuse.
    pub fn new() -> Self {
        let mut log = [0; 256];
        let mut antilog = [0; 256];
        let mut x = 1u8;
        for (i, slot) in antilog[..255].iter_mut().enumerate() {
            log[x as usize] = i as u8;
            *slot = x;
            x = Self::xtime(x);
        }
        antilog[255] = antilog[0];
        Self { log, antilog }
    }

    /// Multiply by 2 in GF(2⁸): left-shift, XOR `0x1D` if the high bit was set.
    ///
    /// Useful both for building tables and for the hand-checkable Q parity row.
    #[inline]
    pub fn xtime(a: u8) -> u8 {
        let hi = a & 0x80;
        let shifted = a << 1;
        if hi != 0 {
            shifted ^ 0x1D
        } else {
            shifted
        }
    }

    /// Addition = subtraction = XOR. Implemented — it is the definition of the field.
    #[inline]
    pub fn add(a: u8, b: u8) -> u8 {
        a ^ b
    }

    /// Field multiply via log/antilog tables.
    #[inline]
    pub fn mul(&self, a: u8, b: u8) -> u8 {
        if a == 0 || b == 0 {
            return 0;
        }
        let i = self.log[a as usize] as u16 + self.log[b as usize] as u16;
        self.antilog[(i % 255) as usize]
    }

    /// Multiplicative inverse of a nonzero element.
    pub fn inv(&self, a: u8) -> Result<u8, ErasureError> {
        if a == 0 {
            return Err(ErasureError::Singular(
                "multiplicative inverse of 0 is undefined".into(),
            ));
        }
        Ok(self.antilog[(255 - self.log[a as usize] as u16) as usize])
    }

    /// `a / b = a · b⁻¹`.
    pub fn div(&self, a: u8, b: u8) -> Result<u8, ErasureError> {
        Ok(self.mul(a, self.inv(b)?))
    }
}

impl Default for Gf256 {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Docs §4: `2·0x20 = 0x40`, `4·0x30 = 0xC0`, `8·0x40 = 0x3A`.
    #[test]
    fn hand_worked_products_from_doc() {
        let gf = Gf256::new();
        assert_eq!(gf.mul(2, 0x20), 0x40);
        assert_eq!(gf.mul(4, 0x30), 0xC0);
        assert_eq!(gf.mul(8, 0x40), 0x3A);
        assert_eq!(Gf256::xtime(0x20), 0x40);
        assert_eq!(Gf256::xtime(0x80), 0x1D); // fold: (0x80<<1) ⊕ 0x11D low byte
    }

    #[test]
    fn mul_inv_round_trip_nonzero() {
        let gf = Gf256::new();
        for a in 1u8..=255 {
            let inv = gf.inv(a).expect("nonzero invertible");
            assert_eq!(gf.mul(a, inv), 1, "a={a:#x}");
        }
    }

    #[test]
    fn inv_of_zero_is_singular() {
        let gf = Gf256::new();
        assert!(matches!(gf.inv(0), Err(ErasureError::Singular(_))));
        assert!(matches!(gf.div(1, 0), Err(ErasureError::Singular(_))));
    }

    #[test]
    fn mul_by_zero_is_zero() {
        let gf = Gf256::new();
        assert_eq!(gf.mul(0, 0xAB), 0);
        assert_eq!(gf.mul(0xAB, 0), 0);
        assert_eq!(gf.div(0, 0xAB).unwrap(), 0);
    }

    #[test]
    fn add_is_xor() {
        assert_eq!(Gf256::add(0x10, 0x20), 0x30);
        assert_eq!(Gf256::add(0x30, 0x30), 0x00);
    }

    #[test]
    fn generator_cycles_all_nonzero() {
        let gf = Gf256::new();
        let mut seen = [false; 256];
        for i in 0..255 {
            let x = gf.antilog[i];
            assert_ne!(x, 0);
            assert!(!seen[x as usize], "duplicate antilog[{i}]={x:#x}");
            seen[x as usize] = true;
            assert_eq!(gf.log[x as usize] as usize, i);
        }
        assert_eq!(gf.antilog[255], 1);
    }
}

File: projects/06-object-store/src/object_store/erasure/gf256.py
"""GF(2⁸) arithmetic — the number system every shard byte lives in.

Ordinary integer arithmetic is useless here: it overflows past 255, and division
does not come back to whole numbers. Erasure coding needs a field that is
**closed over `0..255`**, has a multiplicative inverse for every nonzero element
(so the reconstruction matrix can be inverted), and where addition is its own
inverse. That field is GF(2⁸).

## The three facts that make it work

**Addition is XOR.** `a + b = a - b = a ^ b`. There is no carry, so a parity
shard can be un-mixed with the same operation that mixed it — which is why plain
XOR parity (RAID-5, and LRC's local groups) needs no field theory at all.

**Multiplication is polynomial multiplication modulo `0x11D`.** Treat a byte as
a degree-7 polynomial over GF(2); multiplying two gives up to degree 14, which
no longer fits in a byte, so the result is reduced modulo the irreducible
polynomial `x⁸ + x⁴ + x³ + x² + 1`. That is the AES and Linux RAID-6 convention,
which matters only for interoperability — any irreducible polynomial of degree 8
gives an isomorphic field.

**Every nonzero element is a power of 2.** `2` is a generator: repeatedly
doubling from 1 walks all 255 nonzero elements before returning to 1. So
multiplication becomes addition of exponents — `log` and `antilog` tables turn a
loop of shifts and conditional XORs into two lookups and an add, which is what
makes encoding fast enough to run on every byte.

The tables are checked against the hand-worked products in
`docs/12-how-erasure-coding-works.md` §3–§4. A wrong table does not raise
anything; it silently corrupts every shard it ever touches, and you find out
during a restore.
"""

from __future__ import annotations

from . import Singular

__all__ = ["Gf256"]


class Gf256:
    """GF(2⁸) with the AES / RAID-6 reduction polynomial `0x11D`.

    Build one and reuse it — the tables are 512 bytes and rebuilding them per
    call would dominate the arithmetic they exist to speed up.
    """

    REDUCTION_POLY = 0x11D
    """`x⁸ + x⁴ + x³ + x² + 1`."""

    REDUCTION_BYTE = REDUCTION_POLY & 0xFF
    """The low byte, XOR'd in by `xtime` after an overflow."""

    HIGH_BIT = 0x80
    MUL_ORDER = 255
    """The multiplicative group's order — exponents are taken modulo this."""

    __slots__ = ("antilog", "log")

    def __init__(self) -> None:
        log = [0] * 256
        antilog = [0] * 256
        value = 1
        for exponent in range(self.MUL_ORDER):
            log[value] = exponent
            antilog[exponent] = value
            value = self.xtime(value)
        # `2**255 == 2**0 == 1`, so index 255 mirrors index 0. Having it means
        # `mul` can add two exponents up to 254 each and index without a modulo
        # in the common case.
        antilog[self.MUL_ORDER] = antilog[0]
        self.log = log
        self.antilog = antilog

    @staticmethod
    def xtime(a: int) -> int:
        """Multiply by 2: shift left, and reduce if the high bit overflowed.

        The primitive everything else is built from, and the one operation worth
        being able to do on paper — it is how the tables above get filled and how
        the RAID-6 Q row is hand-checked.
        """
        shifted = (a << 1) & 0xFF
        return shifted ^ Gf256.REDUCTION_BYTE if a & Gf256.HIGH_BIT else shifted

    @staticmethod
    def add(a: int, b: int) -> int:
        """Addition, which is also subtraction, which is XOR."""
        return a ^ b

    def mul(self, a: int, b: int) -> int:
        """Multiply via the log tables."""
        if a == 0 or b == 0:
            # Not a table lookup: `log[0]` is undefined (no exponent gives 0),
            # so zero has to be special-cased rather than encoded.
            return 0
        return self.antilog[(self.log[a] + self.log[b]) % self.MUL_ORDER]

    def inv(self, a: int) -> int:
        """The multiplicative inverse of a nonzero element."""
        if a == 0:
            raise Singular("multiplicative inverse of 0 is undefined")
        return self.antilog[self.MUL_ORDER - self.log[a]]

    def div(self, a: int, b: int) -> int:
        """`a / b = a · b⁻¹`."""
        return self.mul(a, self.inv(b))
